File: Code/C++/Rotation_count.cpp
/*

 C++ program to find number of rotations
 in a sorted and rotated array.


In an given sorted array,we will have to find the number of count
of rotation after which we have got this array.Here, we take into
observation the element i which is greater than the element i+1.
After we get this element we calculate the number of rotation that might have occured.
*/



#include<bits/stdc++.h>
using namespace std;
 
// Returns count of rotations for an array
int countRotations(int arr[], int n)
{
   //here we find index of minimum element
    int min = arr[0], min_index;
    for (int i=0; i<n; i++)
    {
        if (min > arr[i])
        {
            min = arr[i];
            min_index = i;
        }
    }
    return min_index;
}
 
// Driver code
int main()
{

    int arr[20],n,i;
    cout<<"enter the number of element\n";
    cin>>n;
    cout<<"now enter the numbers one by one:\n";
    for(i=0;i<n;i++)
    {
        cin>>arr[i];
    }
    
    cout << countRotations(arr, n);
    return 0;
}

/*

input: {15,16,18,19,1,2,4,6,7}
output:4 //after 4 right rotation of {1,2,4,6,7,15,16,18,19} we get this output

input:{1,2,3,4,5,6}
output:0 //there is no rotation

*/

File: Code/Python/RotationCount.py
'''
Descrption ->
            Given an array of distinct number enteries which are arranged in ascending order.
            If the array is rotated clockwise r number of times. Then, calculate the total 
            number of rotations(r) in the respective array.
'''

# program to find number of rotations in a sorted rotated array. 
  
# PYTHON CODE:
 
def countRotations(arr,low, high):  
    
    # When array in not rotated at all
    if (high < low): 
        return 0
  
    # If there is only one  
    # element left 
    if (high == low): 
        return low 
  
    # Find mid 
    # (low + high)/2  
    mid = low + (high - low)/2;  
    mid = int(mid) 
  
    # Check if element (mid+1) is 
    # minimum element. Consider  
    # the cases like {3, 4, 5, 1, 2} 
    if (mid < high and arr[mid+1] < arr[mid]): 
        return (mid+1) 
  
    # Check if mid itself is  
    # minimum element 
    if (mid > low and arr[mid] < arr[mid - 1]): 
        return mid 
  
    # Decide whether we need to go 
    # to left half or right half 
    if (arr[high] > arr[mid]): 
        return countRotations(arr, low, mid-1); 
  
    return countRotations(arr, mid+1, high) 
  
# Driver code 
arr = [15, 18, 2, 3, 6, 12] 
n = len(arr) 
print(countRotations(arr, 0, n-1))     
